### PR TITLE
[native] Handle NULLIF without the need for specialForm

### PR DESCRIPTION
**Pull request by vermapratyush**
The usage of type coercion in NULLIF in presto/java made it incompatible with 
presto native, this was made possible by specialForm. Handle NULLIF in presto 
native using the expanded form:

> if (equal(cast(first as <common type>), cast(second as <common type>)), cast(null as firstType), first)

Extend TestExpressionCompiler to native test suite with addition of nullIf in 
GeneralQueries test suite.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by rschlussel — COMMENTED** [review]
can't find the actual change.  can you point me to which file it's in?  also, can you add an end to end test to AbstractTestQueries to make sure a full query will work (I think presto native extends/runs all those queries, right? if not, then some equivalent to add a test for both java and for native implementations).

**Review by rschlussel — COMMENTED** [review]
Never mind about not being able to find the main change.  Looks like it was a temporary state while you were updating the branch.

**Comment by rschlussel** [comment]
also fix checkstyle violations

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java [hunk 1]
```diff
@@ -195,6 +195,14 @@ public void tearDown(Method method)
         log.info("FINISHED %s in %s verified %s expressions", method.getName(), Duration.nanosSince(start), futures.size());
     }
 
+    // This is a setter and not a test method.
+    // TestNG considers this as a test method since we have annotated the class with @Test.
+    @Test(enabled = false)
```
[the remark this answers is not in the document]
Comment (vermapratyush) [reply presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java:200]: The field `functionAssertions` is private, hence a no argument `setFunctionAssertions()` wont' be able to set `this.functionAssertions`. Instead of changing the access to public, instead I created the function `setFunctionAssertions(FunctionAssertions functionAssertions)` and calling it from base and native class.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeGeneralQueriesJSON.java [hunk 2]
```diff
@@ -15,19 +15,29 @@
...
+    protected ExpectedQueryRunner createExpectedQueryRunner()
+            throws Exception
     {
         return PrestoNativeQueryRunnerUtils.createJavaQueryRunner();
     }
+
+    // TODO: Fix nullif for presto native
+    @Override
+    @Ignore
+    public void testNullIf()
```
[the remark this answers is not in the document]
Comment (vermapratyush) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeGeneralQueriesJSON.java:40]: This fix is for sapphire/velox integration, although I am checking with presto native members to suggest what could be wrong.

[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeGeneralQueriesJSON.java:40]: thanks. in general if there is a todo for something broken it's good to create a github issue with a description of the problem and link it in the comment.  That way it's easy for someone seeing the todo to understand what's going on with fixing it.

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java:200]: I think I meant make this a method that returns the desired functionAssertion that gets overridden in the native implementation.  e.g. `public FunctionAssertions createFunctionAssertions()`, and then in the beforeClass method do`functionAssertions = createFunctionAssertions()`.  Seems better than making a public setter method.  

Either way, shouldn't need the `@Test` annotation

[the remark this answers is not in the document]
Comment (vermapratyush) [reply presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java:200]: Sounds good.
`public FunctionAssertions createFunctionAssertions` does not require the @Test annotation, I think a public method with void return type was automatically being considered by the framework due to class level annotation @Test.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestExpressionCompiler.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    @Override
+    @Ignore
+    public void testBinaryOperatorsDecimalDouble()
+            throws Exception
+    {
+    }
+
+    // Remove the override from the following tests on a need basis, not all expressions have custom handling for native query runner, hence they are ignored.
```
[the remark this answers is not in the document]
Comment (vermapratyush) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestExpressionCompiler.java:58]: They still pass, but running them seems redundant as there is no custom handling. The queries gets executed within JVM (constant folding) and does not touch the cpp server.

**Comment by vermapratyush** [comment]
The test failure is unrelated to this change, I am looking at fixing the test here. #20003

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -2099,10 +2099,7 @@ private List<FunctionCall> analyzeWindowFunctions(QuerySpecification node, List<
                 }
 
                 Window window = windowFunction.getWindow().get();
-                if (window.getOrderBy().filter(
-                                orderBy -> orderBy.getSortItems().stream().anyMatch(
-                                        item -> item.getSortKey() instanceof Literal))
-                        .isPresent()) {
+                if (window.getOrderBy().filter(orderBy -> orderBy.getSortItems().stream().anyMatch(item -> item.getSortKey() instanceof Literal)).isPresent()) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2102]: unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java [hunk 5]
```diff
@@ -108,6 +108,7 @@ public static RowExpression getSqlFunctionRowExpression(
                         Optional.empty(),
                         sqlFunctionProperties,
                         sessionFunctions,
+                        false,
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:111]: is this safe to assume it's not native?

File: presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java [hunk 6]
```diff
@@ -44,7 +44,7 @@ protected void assertQuery(String sql)
     // TODO: Enable following Ignored tests after fixing (Tests can be enabled by removing the method)
     @Override
     @Ignore
-    public void testCatalogWithCacheEnabled() {}
+    public void testBucketedExecution() {}
```
Comment (rschlussel) [inline presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java:47]: unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (vermapratyush) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2102]: This line had checkstyle violations.

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2102]: oh probably just belongs in the first commit.  looks like when you added the functionAndtypeManager in this class you reformatted this line too.

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java (hunk 5 above)
Comment (vermapratyush) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:111]: I don't see this code path being triggered in native path.
It is used in 
- https://github.com/vermapratyush/presto/blob/d5eed561c31e21cb0b6cfa5f94038d9d4f2b4b89/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineSqlFunctions.java#L106
https://github.com/prestodb/presto/blob/d5eed561c31e21cb0b6cfa5f94038d9d4f2b4b89/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/InlineSqlFunctions.java#L106 - should not be invoked in native flow
- 
- https://github.com/vermapratyush/presto/blob/d5eed561c31e21cb0b6cfa5f94038d9d4f2b4b89/presto-main/src/main/java/com/facebook/presto/sql/gen/RowExpressionCompiler.java#L144
https://github.com/prestodb/presto/blob/d5eed561c31e21cb0b6cfa5f94038d9d4f2b4b89/presto-main/src/main/java/com/facebook/presto/sql/gen/RowExpressionCompiler.java#L144 - should not be invoked in native flow
- https://github.com/vermapratyush/presto/blob/d5eed561c31e21cb0b6cfa5f94038d9d4f2b4b89/presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java#L282
https://github.com/prestodb/presto/blob/d5eed561c31e21cb0b6cfa5f94038d9d4f2b4b89/presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java#L282 - Not sure about this one (i don’t see this getting invoked from the e2e unit tests)
Do you know if RowExpressionInterpreter is of concern?

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:111]: I think these can be invoked from native flow if someone uses a sql function that uses NULLIF in its implementation. And seems like an easy place for bugs to lurk.

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeGeneralQueriesJSON.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vermapratyush) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeGeneralQueriesJSON.java]: This issue is fixed now

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java (hunk 5 above)
Comment (aditi-pandit) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:113]: Could this be because all the native tests set configuration "inline-sql-functions" to false ? https://github.com/prestodb/presto/blob/master/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/NativeQueryRunnerUtils.java#L32

Ideally, we should try to avoid such properties/configurations. 

But is it possible to use that property instead in this rule ?

Comment (vermapratyush) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:113]: @aditi-pandit Can you please elaborate what do you mean by `can we use that property instead of this rule`?

Comment (aditi-pandit) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:113]: @vermapratyush : I'm not super familiar with optimizer code, but it is possible that instead of hardcoding to false, we get the value for inline-sql-functions property at a higher level (or here itself if there is access to it) and use that instead in this code ?

Comment (vermapratyush) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:113]: The issue with `inline-sql-functions` is the same as that of using Session to determine if the query is for native, certain code-flow does not have session object.
I also verified with the velox team, and it would not make sense to provide support for SQL function for native queries within the java stack. It is better to support sql function by implementing it natively in velox itself. So I think hard-coding is ok here.

**Review by rschlussel — APPROVED** [review]
Summarizing discussion:
This fixes handling of NULLIF for native for all code paths that native queries currently go through.  The SqlFunctionUtils, which are used in the sql function code path does not currently properly support native, since there is no access to the session there, only connector session.  Sqlfunctions are disabled in native, so this path isn't currently accessed, but it's a place for surprise bugs for the future. This is still a todo to fix, but we don't want to block the main issue on addressing this case: https://github.com/prestodb/presto/issues/20008.

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java [hunk 7]
```diff
@@ -108,6 +108,9 @@ public static RowExpression getSqlFunctionRowExpression(
                         Optional.empty(),
                         sqlFunctionProperties,
                         sessionFunctions,
+                        // TODO: use session to determine if this is a native query
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:111]: Why can't we address this TODO now?

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestExpressionCompiler.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    @Override
+    public void setupClass()
+    {
+        super.setupClass();
+        setFunctionAssertions(new FunctionAssertions(getQueryRunner().getDefaultSession(), new FeaturesConfig().setNativeExecutionEnabled(true)));
+    }
+
+    protected abstract QueryRunner getQueryRunner();
+
+    // TODO: The following test have trouble converting long to Decimal.
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestExpressionCompiler.java:33]: @majetideepak Deepak, is this something you could help look into?

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java (hunk 7 above)
Comment (vermapratyush) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:111]: Detailed conversation in this thread: https://github.com/prestodb/presto/pull/19791/files/189afe0a571d7ee128f242fd803a6098d967861b#r1243767843

In short,
> The SqlFunctionUtils, which are used in the sql function code path does not currently properly support native, since there is no access to the session there, only connector session.

**Review by mbasmanova — COMMENTED** [review]
@vermapratyush There was custom translation of NULLIF expression in Prestissimo. I assume we do not need this anymore. Let's remove.

**Review by mbasmanova — APPROVED** [review]
@rschlussel @vermapratyush Do you have a sense of any performance penalty this implementation of NULLIF would have?

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java (hunk 7 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java:111]: the solution would be to add the isNative flag to the connector session and pass it through when we create the connector session.  I think that's a very reasonable to do, and there may be other places that would want to handle things differently depending on the native flag too, so could be generally useful.

**Comment by rschlussel** [comment]
> @rschlussel @vermapratyush Do you have a sense of any performance penalty this implementation of NULLIF would have?

it's doing the same translation that the custom cpp code is doing, so should be the same. Only overhead would be if it's adding casts that the cpp function doesn't need (e.g. type only coercion)

File: presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java [hunk 9]
```diff
@@ -41,6 +42,21 @@ protected void assertQuery(String sql)
         PrestoSparkNativeQueryRunnerUtils.assertShuffleMetadata();
     }
 
+    @Test
+    public void testNullIf()
+    {
+        assertQuery("SELECT NULLIF(totalprice, 0) FROM (SELECT SUM(extendedprice) AS totalprice FROM lineitem WHERE shipdate >= '1995-09-01')");
```
Comment (mbasmanova) [inline presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java:48]: These tests look very similar to me. What's the motivation for adding so many nearly identical tests?

File: presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java [hunk 10]
```diff
@@ -41,6 +42,21 @@ protected void assertQuery(String sql)
...
+        assertQuery("SELECT NULLIF(totalprice, 0) FROM (SELECT SUM(extendedprice) AS totalprice FROM lineitem WHERE shipdate >= '9999-99-99')");
+        assertQuery("SELECT NULLIF(totalprice, 0.5) FROM (SELECT SUM(extendedprice) AS totalprice FROM lineitem WHERE shipdate >= '1995-09-01')");
+        assertQuery("SELECT NULLIF(totalprice, 0.5) FROM (SELECT SUM(extendedprice) AS totalprice FROM lineitem WHERE shipdate >= '9999-99-99')");
+        assertQuery("SELECT NULLIF(totalprice, 0) FROM (SELECT COUNT(1) AS totalprice FROM lineitem WHERE shipdate >= '1995-09-01')");
+        assertQuery("SELECT NULLIF(totalprice, 0) FROM (SELECT COUNT(1) AS totalprice FROM lineitem WHERE shipdate >= '9999-99-99')");
+        assertQuery("SELECT NULLIF(totalprice, 0.5) FROM (SELECT COUNT(1) AS totalprice FROM lineitem WHERE shipdate >= '1995-09-01')");
+        assertQuery("SELECT NULLIF(totalprice, 0.5) FROM (SELECT COUNT(1) AS totalprice FROM lineitem WHERE shipdate >= '9999-99-99')");
+        assertQuery("SELECT NULLIF('abcd', 'abcd')");
```
Comment (mbasmanova) [inline presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java:56]: This runs on coordinator, hence, doesn't test any C++ code.

File: presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java (hunk 9 above)
Comment (vermapratyush) [reply presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeGeneralQueries.java:48]: Permutation combination of bigint, double and empty rows (see date filter).

File: presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestExpressionCompiler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (majetideepak) [reply presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestExpressionCompiler.java:33]: @mbasmanova sure.
